scigym/envs/surfacecode/surfacecode_env.py

```diff
@@ -0,0 +1,717 @@
+
+	def __init__(self, p_phys=0.01, p_meas=0.01, error_model="DP", use_Y=True, volume_depth=3):
+
+		self.d = 5
+		self.p_phys = p_phys
+		self.p_meas= p_meas
+		self.error_model = error_model
+		self.use_Y = use_Y
+		self.volume_depth = volume_depth
+
+		self.n_action_layers = 0
+		if error_model == "X":
+			self.num_actions = self.d**2 + 1
+			self.n_action_layers = 1
+			static_decoder_path = pkg_resources.resource_filename('scigym', 'envs/surfacecode/referee_decoders/X_decoder')
+		elif error_model == "DP":
+			static_decoder_path = pkg_resources.resource_filename('scigym', 'envs/surfacecode/referee_decoders/X_decoder')
+			if use_Y:
+				self.num_actions = 3*self.d**2 + 1
+				self.n_action_layers = 3
+			else:
+				self.num_actions = 2*self.d**2 + 1 
+				self.n_action_layers = 2
+		else:
+			print("specified error model not currently supported!")
+
+		self.static_decoder = load_model(static_decoder_path)
+
+		self.identity_index = self.num_actions -1
+		self.identity_indicator = self.generate_identity_indicator(self.d)
+
+		self.qubits = self.generateSurfaceCodeLattice(self.d)
+		self.qubit_stabilizers = self.get_stabilizer_list(self.qubits, self.d)  
+		self.qubit_neighbours = self.get_qubit_neighbour_list(self.d) 
+		self.completed_actions = np.zeros(self.num_actions, int)
+		
+	
+		self.observation_space=gym.spaces.Box(low=0,high=1,
+									  shape=(self.volume_depth+self.n_action_layers,
+										2*self.d+1, 
+										2*self.d+1),
+									  dtype=np.uint8)
+		
+		self.action_space = gym.spaces.Discrete(self.num_actions)
+
+		self.hidden_state = np.zeros((self.d, self.d), int)
+		self.current_true_syndrome = np.zeros((self.d+1, self.d+1), int)
+		self.summed_syndrome_volume = None         
+		self.board_state = np.zeros((self.volume_depth + self.n_action_layers, 2 * self.d + 1, 2 * self.d + 1),int)
+
+		self.completed_actions = np.zeros(self.num_actions, int)
+		self.acted_on_qubits = set()
+		self.legal_actions = set()
+		self.done = False
+		self.lifetime = 0
+
+		self.multi_cycle = True
+
+	def reset(self):
+		"""
+		Resetting of the environment introduces a new non-trivial syndrome volume.
+
+		:return: self.board_state: The new reset visible state of the environment = syndrome volume + blank action history volume
+		"""
+
+		self.done = False
+		self.lifetime = 0
+		
+		# Create the initial error - we wait until there is a non-trivial syndrome. BUT, the lifetime is still updated!
+		self.initialize_state()
+
+		# Update the legal moves available to us
+		self.reset_legal_moves()
+
+		return self.board_state
+
+	def step(self, action):
+		"""
+		Given an action, this method executes the logic of the environment.
+
+		:param: action: The action, given as an integer, supplied by some agent.
+		:return: self.board_state: The new reset visible state of the environment = syndrome volume + action history volume
+		:return: reward: The reward for the action
+		:return: self.done: The boolean terminal state indicator
```

Comment:
I tested the basic functionality in your code, and I was wondering what the `done` value indicates in your case? I am asking because in the few trials I did, I always received a `done=True`, immediately.

Reply:
The `done` value is a boolean which indicates whether or not the current env state is a terminal state. In this environment, the value of `done` is determined by whether or not a suitable referee decoder can successfully decode the perfect syndrome corresponding to the underlying error configuration.

I ran some trials, and indeed for the default environment settings (as they were) after a single action the value of `done` was almost always true. This is because when `env.reset()` is run, the error channel is applied until there is a non-trivial syndrome - if the effective error rate is too high (taking into account the initial `volume_depth`) then resetting the environment will almost always result in an error configuration whose perfect syndrome cannot be correctly decoded by the referee decoder.

I changed the default environment settings to a more realistic error rate, error model and volume depth, and now it is possible to confirm that in general resetting the environment leads to error configurations with a small number of errors, on top of which a few actions can normally be performed before a terminal state is reached.

In essence, the default "difficulty level" of the game was set too high, and this has been changed.